src/modelutils.py

```diff
@@ -24,32 +26,56 @@ def skip(*args, **kwargs):
         torch.nn.init.kaiming_uniform_, torch.nn.init.uniform_, torch.nn.init.normal_ = saved_inits  # restoring
 
 
-def get_model(model_path, load_quantized=None, dtype="auto", model_seqlen=2048):
+def dispatch_quantized_model(model):
+    num_devices = torch.cuda.device_count()
+    device_map = {"model.embed_tokens": 0, "model.norm": num_devices - 1, "lm_head": num_devices - 1}
+    num_layers = len(get_layers(model))
+    layers_per_device = math.ceil(num_layers / num_devices)
+    for layer_id in range(num_layers):
+        device_id = layer_id // layers_per_device
+        device_map[f"model.layers.{layer_id}"] = device_id
+    model = dispatch_model(model, device_map)
+    return model
+
+
+def get_model(
+    model_path, load_quantized=None, dtype="auto", model_seqlen=2048, device_map=None, attn_implementation=None
+):
     if dtype == "auto":
         dtype = (
             AutoConfig.from_pretrained(model_path, trust_remote_code=True).torch_dtype or "auto"
         )  # force transformers 4.29.2 to follow the same rules as 4.30.x
     else:
         dtype = getattr(torch, dtype)
 
+    model_kwargs = {}
+    # this argument is avaialbe only for transformers >= 4.38.0
+    if transformers.__version__ >= "4.38.0":
+        model_kwargs["attn_implementation"] = attn_implementation
+
     with suspend_nn_inits():
+        model = AutoModelForCausalLM.from_pretrained(
+            pretrained_model_name_or_path=model_path,
+            trust_remote_code=True,
+            torch_dtype=dtype,
+            # defer distribution if loading quantized
+            device_map=None if load_quantized else device_map,
+            low_cpu_mem_usage=True,
+            local_files_only=True,
+            **model_kwargs,
+        )
         if load_quantized:
             print("Initializing model with random weights...")
-            config = AutoConfig.from_pretrained(model_path, trust_remote_code=True)  # consider trust_remote_code=True
-            model = AutoModelForCausalLM.from_config(config, trust_remote_code=True, torch_dtype=dtype).eval()
             print("Loading quantized model ...")
             model = load_quantized_model(model, load_quantized)
+            # TODO works only for Llama
```

Comment:
if still true, please make this into an assert statement (e.g. assert config["model_type"] == "llama")